syntax = "proto3";

package booksuggest;

import "google/protobuf/timestamp.proto";

service BookSuggestion {
    rpc InitializeOrder (SuggestionInitRequest) returns (SuggestionInitResponse);
    rpc GetSuggestions (RecommendationRequest) returns (RecommendationResponse);
    rpc ClearOrder(ClearOrderRequest) returns (ClearOrderResponse);
}

message SuggestionInitRequest {
  string order_id = 1;
  repeated string book_tokens = 2;
  int32 limit = 3;
  map<string, int32> vectorClock = 4; 
}

message SuggestionInitResponse {
  bool success = 1;
  map<string, int32> vectorClock = 2; 
}

message RecommendationRequest {
  string order_id = 1; 
  map<string, int32> vectorClock = 2; 
}

message RecommendationResponse {
  repeated Recommendation recommendations = 1;
  map<string, int32> vectorClock = 2; 
}

message Recommendation {
  Book book = 1;
  float confidence_score = 2;
  string reason = 3;
}

message Book {
  string id = 1;
  string title = 2;
  string author = 3;
  string description = 4;
  repeated string genres = 6;
  google.protobuf.Timestamp publish_date = 7;
}

message ClearOrderRequest {
    string order_id = 1;
    map<string, int32> vectorClock = 2;
}

message ClearOrderResponse {
    bool success = 1;
    string error = 2;
    map<string, int32> vectorClock = 3;
}
